Try not to use verbatim paths in `Command::current_dir`

ChrisDenton opened the pull request:
If possible, we should try not to use verbatim paths in `Command::current_dir`. It might work but it might also break code in the subprocess that assume the current directory isn't verbatim (including Windows APIs). cc @ehuss

Side note: we now have a lot of ad-hoc fixes like this spread about the place. It'd be good to make a proper `WindowsPath` type that handles all this in one place. But that's a bigger job for another PR.

ChrisDenton committed:
Test Command::current_dir with verbatim path

Noratrieb commented on tests/ui/process/win-command-curdir-no-verbatim.rs:
```diff
@@ -0,0 +1,35 @@
+// Test that windows `creation_flags` extension to `Command` works.
```
this is wrong

ChrisDenton replied:
Oh, lol. I copy/pasted the template and forgot to change the comment...

Noratrieb reviewed (commented):
this looks reasonable to me, but I'd like someone with more clues about Windows to take a look as well. I don't know who'd be a great fit, feel free to assign whoever you want

ChrisDenton committed:
Avoid verbatim paths in Command::current_dir

If possible, we should try not to use verbatim paths in Command::current_dir. It might work but it might also break code (including some Windows APIs) that assume the current directory isn't verbatim.

ChrisDenton commented:
> I'd like someone with more clues about Windows to take a look as well

I'm honestly not sure, so choosing someone at random:
r? @tgross35
Want to take a stab at this? Feel free to pass.

tgross35 commented on library/std/src/sys/path/windows.rs:
```diff
@@ -350,3 +350,45 @@ pub(crate) fn absolute(path: &Path) -> io::Result<PathBuf> {
+    if result == 0 || result as usize != buffer_len - 1 {
+        false
+    } else {
+        unsafe {
+            new_path.set_len((result as usize) + 1);
+        }
```
```rust
// SAFETY: `GetFullPathNameW` initialized `result` bytes and does not exceed `nBufferLength - 1` (capacity).
```

or similar

tgross35 commented on library/std/src/sys/path/windows.rs:
```diff
@@ -350,3 +350,45 @@ pub(crate) fn absolute(path: &Path) -> io::Result<PathBuf> {
     path.has_root() && path.prefix().is_some()
 }
+
+/// Test that the path is absolute, fully qualified and unchanged when processed by the Windows API.
+///
+/// For example:
+///
+/// - `C:\path\to\file` will return true.
+/// - `C:\path\to\nul` returns false because the Windows API will convert it to \\.\NUL
+/// - `C:\path\to\..\file` returns false because it will be resolved to `C:\path\file`.
+///
+/// This is a useful property because it means the path can be converted from and to and verbatim
+/// path just by changing the prefix.
```
Verbatim paths can contain special names like COM while regular paths cannot right? That may be worth mentioning and adding a test for.

tgross35 commented on library/std/src/sys/process/windows.rs:
```diff
@@ -880,9 +880,33 @@ fn make_envp(maybe_env: Option<BTreeMap<EnvKey, OsString>>) -> io::Result<(*mut
+                let start = r"\\?\UN".len();
+                dir_str[start] = b'\\' as u16;
+                if path::is_absolute_exact(&dir_str[start..]) {
+                    unsafe { dir_str.as_ptr().add(start) }
```
Could this be `&dir_str[start..].as_ptr()` to avoid the unsafe? Applies once below too.

tgross35 reviewed (commented):
I am also not a Windows expert but it looks pretty reasonable to me. Sorry for the delay, just a few nits/questions.

tgross35 reviewed (approved):
LGTM, r=me with the last commit squashed

ChrisDenton commented:
@bors r=tgross35

bors commented:
:pushpin: Commit edfc74722556c659de6fa03b23af3b9c8ceb8ac2 has been approved by `tgross35`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=tgross35 edfc74722556c659de6fa03b23af3b9c8ceb8ac2 -->
<!-- homu: {"type":"Approved","sha":"edfc74722556c659de6fa03b23af3b9c8ceb8ac2","approver":"tgross35","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
